Mathlib/Algebra/Order/Antidiag/Nat.lean

```diff
@@ -0,0 +1,247 @@
+/-
+Copyright (c) 2024 Arend Mellendijk. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Arend Mellendijk
+-/
+
+import Mathlib.Algebra.Order.Antidiag.Pi
+import Mathlib.NumberTheory.ArithmeticFunction
+
+/-!
+# Sets of tuples with a fixed product
+
+This file defines the finite set of `d`-tuples of natural numbers with a fixed product `n` as
+`Nat.finMulAntidiagonal`.
+
+## Main Results
+* There are `d^(ω n)` ways to write `n` as a product of `d` natural numbers, when `n` is squarefree
+(`card_finMulAntidiagonal_of_squarefree`)
+-/
+
+open Finset
+open scoped BigOperators ArithmeticFunction
+namespace PNat
+
+instance instHasAntidiagonal : Finset.HasAntidiagonal (Additive ℕ+) :=
+  /- The set of divisors of a positive natural number.
+This is `Nat.divisorsAntidiagonal` without a special case for `n = 0`. -/
+  let divisorsAntidiagonal (n : ℕ+) : Finset (ℕ+ × ℕ+) :=
+    (Nat.divisorsAntidiagonal n).attach.map
+      ⟨fun x =>
+        (⟨x.val.1, Nat.pos_of_mem_divisors <| Nat.fst_mem_divisors_of_mem_antidiagonal x.prop⟩,
+        ⟨x.val.2, Nat.pos_of_mem_divisors <| Nat.snd_mem_divisors_of_mem_antidiagonal x.prop⟩),
+      fun _ _ h => Subtype.ext <| Prod.ext (congr_arg (·.1.val) h) (congr_arg (·.2.val) h)⟩
+
+  have mem_divisorsAntidiagonal {n : ℕ+} (x : ℕ+ × ℕ+) :
+    x ∈ divisorsAntidiagonal n ↔ x.1 * x.2 = n := by
+    simp_rw [divisorsAntidiagonal, Finset.mem_map, Finset.mem_attach, Function.Embedding.coeFn_mk,
+      Prod.ext_iff, true_and, ← coe_inj, Subtype.exists]
+    aesop
+  { antidiagonal := fun n ↦ divisorsAntidiagonal (Additive.toMul n) |>.map
+      (.prodMap (Additive.ofMul.toEmbedding) (Additive.ofMul.toEmbedding))
+    mem_antidiagonal := fun {n a} ↦ by
+      obtain ⟨a₁, a₂⟩ := a
+      revert n a₁ a₂
+      simp_rw [Additive.ofMul.surjective.forall, ← ofMul_mul, Additive.ofMul.injective.eq_iff]
+      simp [mem_divisorsAntidiagonal] }
```

Comment:
```suggestion
    mem_antidiagonal := by simp [← ofMul_mul, mem_divisorsAntidiagonal] }
```
should work, but I couldn't test it because the cache on your branch doesn't work for some reason